### DISCUSSION TITLE
Caching pip dependencies and mounting secret

### DISCUSSION THREAD

**Discussion by V0lantis** [apache/airflow#22624] [Q&A] [answered]
I am not sure if this has already been discussed somewhere else, so I am first opening a discussion, but night open an issue in the future:

1. Is it planned to add the the [_pip docker cache_ ](https://pythonspeed.com/articles/docker-cache-pip-downloads/) available with `DOCKER_BUILDKIT`? It would speed up our custom image build in my company
2. Is it planned to use the [secret mounting](https://docs.docker.com/develop/develop-images/build_enhancements/#new-docker-build-secret-information)?

Unfortunately, the later point is a security concern we have, since we would just need to download one of the build layer to access our credentials through [`pip.conf`](https://github.com/apache/airflow/blob/8240a447ae8296d904b5661a9bff182f4aa55d06/Dockerfile#L1254) file.

Thank you so much for any kind of helps for those to questions

**Comment by potiuk** [comment, marked as answer]
Some good and bad news.

* Airflow Image is already using buildkit in 2.3.0 (even requires buildkit) -> See the [Changelog](https://github.com/apache/airflow/blob/main/docs/docker-stack/changelog.rst) that is describing it (will be released in 2.3.0)
* I am working on improving caching in both prod and cI image. And adding --cache mount in PROD image's build layer will be next I think.
* I will have a look at using secret as well, but I am not sure we need them. For now you can actually easily use pip.conf and keep it secret. See https://airflow.apache.org/docs/docker-stack/build.html#using-custom-installation-sources  - the pip.conf that you will place in docker-context-files will only be stored in "build" segment duiring the build and this segment is only used to copy the installed python code, nothing else, so your pip.conf will never get to the final image. However if you think you can improve this - PRs are most welcome.

**Reply by V0lantis** [reply]
Thank you so much @potiuk as always !

> * I will have a look at using secret as well, but I am not sure we need them. For now you can actually easily use pip.conf and keep it secret. See https://airflow.apache.org/docs/docker-stack/build.html#using-custom-installation-sources  - the pip.conf that you will place in docker-context-files will only be stored in "build" segment duiring the build and this segment is only used to copy the installed python code, nothing else, so your pip.conf will never get to the final image. However if you think you can improve this - PRs are most welcome.

 I'll try to have a look at it even though I am not sure it is easily doable.
